File: apps/api/app/services/oshi_contrast.py
"""推し色のコントラスト選定（WCAG AA 本文 4.5:1 目安）。

Web `oshiContrast.ts` と同ルール。文字色はユーザー選択せず自動。
"""

from __future__ import annotations

import re
from typing import TypedDict

HEX_RE = re.compile(r"^#([0-9A-Fa-f]{6})$")
AA_NORMAL = 4.5

# 固定候補（白／濃色〜95%相当の読みやすいほぼ黒）
FG_WHITE = "#ffffff"
FG_NEAR_BLACK = "#1a1614"

DEFAULT_MAIN_HEX = "#9f606c"
DEFAULT_SUB_HEX = "#6a9bb8"


class ResolvedOshiColors(TypedDict):
    main_hex: str
    sub_hex: str
    main_foreground: str
    soft_bg: str
    soft_foreground: str


def normalize_hex(raw: str) -> str:
    value = (raw or "").strip()
    if not value.startswith("#") and len(value) == 6:
        value = f"#{value}"
    m = HEX_RE.match(value)
    if not m:
        raise ValueError("色の形式が不正です（#RRGGBB）")
    return f"#{m.group(1).lower()}"


def _parse_rgb(hex_color: str) -> tuple[int, int, int]:
    h = normalize_hex(hex_color)[1:]
    return int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)


def _to_hex(r: int, g: int, b: int) -> str:
    return f"#{r:02x}{g:02x}{b:02x}"


def _channel_to_linear(c: int) -> float:
    s = c / 255.0
    return s / 12.92 if s <= 0.03928 else ((s + 0.055) / 1.055) ** 2.4


def relative_luminance(hex_color: str) -> float:
    r, g, b = _parse_rgb(hex_color)
    return (
        0.2126 * _channel_to_linear(r)
        + 0.7152 * _channel_to_linear(g)
        + 0.0722 * _channel_to_linear(b)
    )


def contrast_ratio(hex_a: str, hex_b: str) -> float:
    l1 = relative_luminance(hex_a)
    l2 = relative_luminance(hex_b)
    lighter = max(l1, l2)
    darker = min(l1, l2)
    return (lighter + 0.05) / (darker + 0.05)


def mix_hex(hex_a: str, hex_b: str, amount: float) -> str:
    """amount=0 で A、1 で B。"""
    t = max(0.0, min(1.0, amount))
    ar, ag, ab = _parse_rgb(hex_a)
    br, bg, bb = _parse_rgb(hex_b)
    return _to_hex(
        int(round(ar + (br - ar) * t)),
        int(round(ag + (bg - ag) * t)),
        int(round(ab + (bb - ab) * t)),
    )


def _fg_candidates(bg_hex: str) -> list[str]:
    light = mix_hex(bg_hex, FG_WHITE, 0.45)
    dark = mix_hex(bg_hex, FG_NEAR_BLACK, 0.55)
    # 重複除去（順序維持）
    out: list[str] = []
    for c in (FG_WHITE, FG_NEAR_BLACK, light, dark):
        if c not in out:
            out.append(c)
    return out


def best_foreground(bg_hex: str, *, min_ratio: float = AA_NORMAL) -> str:
    bg = normalize_hex(bg_hex)
    scored: list[tuple[float, str]] = []
    for fg in _fg_candidates(bg):
        ratio = contrast_ratio(fg, bg)
        scored.append((ratio, fg))
    scored.sort(key=lambda x: x[0], reverse=True)
    for ratio, fg in scored:
        if ratio >= min_ratio:
            return fg
    # 満たせない場合は最良でも拒否
    raise ValueError("この色では十分なコントラストの文字色を選べません")


def soft_surface(sub_hex: str) -> str:
    """サブ色のやわらかい面（白寄りミックス）。"""
    sub = normalize_hex(sub_hex)
    return mix_hex(FG_WHITE, sub, 0.22)


def resolve_oshi_colors(main_hex: str, sub_hex: str) -> ResolvedOshiColors:
    main = normalize_hex(main_hex)
    sub = normalize_hex(sub_hex)
    main_fg = best_foreground(main)
    soft = soft_surface(sub)
    soft_fg = best_foreground(soft)
    return {
        "main_hex": main,
        "sub_hex": sub,
        "main_foreground": main_fg,
        "soft_bg": soft,
        "soft_foreground": soft_fg,
    }

File: apps/web/src/lib/oshiContrast.ts
/**
 * 推し色のコントラスト選定（WCAG AA 本文 4.5:1 目安）。
 * API `oshi_contrast.py` と同ルール。
 */

export const AA_NORMAL = 4.5;
export const FG_WHITE = "#ffffff";
export const FG_NEAR_BLACK = "#1a1614";
export const DEFAULT_MAIN_HEX = "#9f606c";
export const DEFAULT_SUB_HEX = "#6a9bb8";

const HEX_RE = /^#([0-9a-fA-F]{6})$/;

export type ResolvedOshiColors = {
  main_hex: string;
  sub_hex: string;
  main_foreground: string;
  soft_bg: string;
  soft_foreground: string;
};

export function normalizeHex(raw: string): string {
  let value = (raw || "").trim();
  if (!value.startsWith("#") && value.length === 6) {
    value = `#${value}`;
  }
  const m = HEX_RE.exec(value);
  if (!m) {
    throw new Error("色の形式が不正です（#RRGGBB）");
  }
  return `#${m[1].toLowerCase()}`;
}

function parseRgb(hexColor: string): { r: number; g: number; b: number } {
  const h = normalizeHex(hexColor).slice(1);
  return {
    r: Number.parseInt(h.slice(0, 2), 16),
    g: Number.parseInt(h.slice(2, 4), 16),
    b: Number.parseInt(h.slice(4, 6), 16),
  };
}

function toHex(r: number, g: number, b: number): string {
  const clamp = (n: number) => Math.max(0, Math.min(255, Math.round(n)));
  return `#${[clamp(r), clamp(g), clamp(b)]
    .map((n) => n.toString(16).padStart(2, "0"))
    .join("")}`;
}

function channelToLinear(c: number): number {
  const s = c / 255;
  return s <= 0.03928 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4;
}

export function relativeLuminance(hexColor: string): number {
  const { r, g, b } = parseRgb(hexColor);
  return (
    0.2126 * channelToLinear(r) +
    0.7152 * channelToLinear(g) +
    0.0722 * channelToLinear(b)
  );
}

export function contrastRatio(hexA: string, hexB: string): number {
  const l1 = relativeLuminance(hexA);
  const l2 = relativeLuminance(hexB);
  const lighter = Math.max(l1, l2);
  const darker = Math.min(l1, l2);
  return (lighter + 0.05) / (darker + 0.05);
}

export function mixHex(hexA: string, hexB: string, amount: number): string {
  const t = Math.max(0, Math.min(1, amount));
  const a = parseRgb(hexA);
  const b = parseRgb(hexB);
  return toHex(
    a.r + (b.r - a.r) * t,
    a.g + (b.g - a.g) * t,
    a.b + (b.b - a.b) * t,
  );
}

function fgCandidates(bgHex: string): string[] {
  const light = mixHex(bgHex, FG_WHITE, 0.45);
  const dark = mixHex(bgHex, FG_NEAR_BLACK, 0.55);
  const out: string[] = [];
  for (const c of [FG_WHITE, FG_NEAR_BLACK, light, dark]) {
    if (!out.includes(c)) out.push(c);
  }
  return out;
}

export function bestForeground(
  bgHex: string,
  minRatio: number = AA_NORMAL,
): string {
  const bg = normalizeHex(bgHex);
  const scored = fgCandidates(bg)
    .map((fg) => ({ fg, ratio: contrastRatio(fg, bg) }))
    .sort((a, b) => b.ratio - a.ratio);
  for (const { fg, ratio } of scored) {
    if (ratio >= minRatio) return fg;
  }
  throw new Error("この色では十分なコントラストの文字色を選べません");
}

export function softSurface(subHex: string): string {
  return mixHex(FG_WHITE, normalizeHex(subHex), 0.22);
}

export function resolveOshiColors(
  mainHex: string,
  subHex: string,
): ResolvedOshiColors {
  const main_hex = normalizeHex(mainHex);
  const sub_hex = normalizeHex(subHex);
  const soft_bg = softSurface(sub_hex);
  return {
    main_hex,
    sub_hex,
    main_foreground: bestForeground(main_hex),
    soft_bg,
    soft_foreground: bestForeground(soft_bg),
  };
}

/** キュレート済みスウォッチ（Lab ベース＋少し追加） */
export const OSHI_SWATCH_PRESETS = [
  { hex: "#9f606c", labelKey: "rose" },
  { hex: "#d4786a", labelKey: "coral" },
  { hex: "#6a9bb8", labelKey: "sky" },
  { hex: "#9a7eb8", labelKey: "lilac" },
  { hex: "#6a9f8e", labelKey: "mint" },
  { hex: "#c4a35a", labelKey: "gold" },
  { hex: "#e879a9", labelKey: "pink" },
  { hex: "#5b7c99", labelKey: "slate" },
] as const;

export type OshiSwatchLabelKey = (typeof OSHI_SWATCH_PRESETS)[number]["labelKey"];

export function tryResolveOshiColors(
  mainHex: string,
  subHex: string,
): ResolvedOshiColors | null {
  try {
    return resolveOshiColors(mainHex, subHex);
  } catch {
    return null;
  }
}
